#!/bin/bash
python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z0__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z0__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z1__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z1__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z2__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z2__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z3__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z3__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z4__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z4__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z5__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z5__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z6__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z6__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z7__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-11_02-13-41/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z7__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z0__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z0__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z1__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z1__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z2__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z2__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z3__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z3__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z4__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z4__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z5__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z5__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z6__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z6__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

python /home/rodrigo/data/doctorado/repos/GWAS_pipeline/utils/summary-gwas-imputation/src/gwas_parsing.py -gwas_file /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z7__test_std_covariates_PC__GBR_unrelated__qc.tsv -output /media/rodrigo/bd2fea02-a909-49bc-9c6c-e0750941ca71/data/GWAS/output/coma/2020-09-30_12-36-48/test_std_covariates_PC__GBR_unrelated__qc/BGEN/GWAS__z7__harmonized__test_std_covariates_PC__GBR_unrelated__qc.tsv -liftover /home/rodrigo/01_Repos/GWAS_pipeline/data/hg19ToHg38.over.chain.gz -output_column_map SNP variant_id -output_column_map a_1 effect_allele -output_column_map a_0 non_effect_allele -output_column_map BP position -output_column_map CHR chromosome --chromosome_format -output_column_map af frequency -output_column_map P pvalue -output_column_map BETA effect_size -output_column_map SE standard_error -output_order variant_id panel_variant_id chromosome position effect_allele non_effect_allele frequency pvalue zscore effect_size standard_error

